crypto/jwx.go

```diff
@@ -152,6 +179,51 @@ func ParseJWT(tokenString string, f PublicKeyFunc, options ...jwt.ParseOption) (
 	return jwt.ParseString(tokenString, options...)
 }
 
+// ParseJWS parses JWS a object, validates and verifies it.
+func ParseJWS(payload []byte, f PublicKeyFunc) (map[string]interface{}, error) {
+	message, err := jws.Parse(payload)
+	if err != nil {
+		return nil, err
+	}
+	headers, body, _, err := jws.SplitCompact(payload)
+	if err != nil {
+		return nil, err
+	}
+	signatures := message.Signatures()
+	for i := range signatures {
+		signature := signatures[i]
+		// Get and check the algorithm
+		alg := signature.ProtectedHeaders().Algorithm()
+		if !isAlgorithmSupported(alg) {
+			return nil, fmt.Errorf("token signing algorithm is not supported: %s", alg)
+		}
+		// Get the verifier for the algorithm
+		verifier, err := jws.NewVerifier(alg)
+		if err != nil {
+			return nil, err
+		}
+		// Get the key id, and get the associated key
+		kid := signature.ProtectedHeaders().KeyID()
+		key, err := f(kid)
+		if err != nil {
+			return nil, err
+		}
+		// This seems an awkward way of appending 3 arrays.
+		payload := append(append(headers, "."...), body...)
+		err = verifier.Verify(payload, signature.Signature(), key)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var rv = make(map[string]interface{})
+	if err = json.Unmarshal(message.Payload(), &rv); err != nil {
```

Comment:
So the `ParseJWS` method is not the reverse of SignJWS since the latest can use any payload but this one is restricted to JSON?

Reply:
Makes sense, I copy & pasted the ParseJWT method.

Reply:
Fixed it, make the method return an []byte